src/dstack/_internal/server/background/tasks/process_pools.py

```diff
@@ -0,0 +1,181 @@
+            if instance.status in (InstanceStatus.CREATING, InstanceStatus.STARTING):
+                instance.status = InstanceStatus.READY
+                await session.commit()
+        else:
+            if instance.status in (InstanceStatus.READY, InstanceStatus.BUSY):
+                instance.status = InstanceStatus.FAILED
+                await session.commit()
+
+
+@runner_ssh_tunnel(ports=[client.REMOTE_SHIM_PORT], retries=1)  # type: ignore[misc]
+def instance_healthcheck(*, ports: Dict[int, int]) -> bool:
+    shim_client = client.ShimClient(port=ports[client.REMOTE_SHIM_PORT])
+    resp = shim_client.healthcheck()
+    if resp is None:
+        return False  # shim is not available yet
+    return bool(resp.service == "dstack-shim")
+
+
+async def terminate(instance_id: UUID) -> None:
+    async with get_session_ctx() as session:
+        instance = (
+            await session.scalars(
+                select(InstanceModel)
+                .where(InstanceModel.id == instance_id)
+                .options(joinedload(InstanceModel.project))
+            )
+        ).one()
+
+        # TODO: need lock
+
+        jpd = parse_raw_as(JobProvisioningData, instance.job_provisioning_data)
+        BACKEND_TYPE = jpd.backend
+        backends = await backends_services.get_project_backends(project=instance.project)
+        backend = next((b for b in backends if b.TYPE in BACKEND_TYPE), None)
+        if backend is None:
+            raise ValueError(f"there is no backned {BACKEND_TYPE}")
+
+        await run_async(
+            backend.compute().terminate_instance, jpd.instance_id, jpd.region, jpd.backend_data
+        )
+
+        instance.deleted = True
+        instance.deleted_at = get_current_datetime()
+        instance.finished_at = get_current_datetime()
+        instance.status = InstanceStatus.TERMINATED
+
+        logger.info("instance %s terminated", instance.name)
+
+        await session.commit()
+
+
+async def terminate_idle_instance() -> None:
+    async with get_session_ctx() as session:
+        res = await session.execute(
+            select(InstanceModel)
+            .where(
+                InstanceModel.termination_policy == TerminationPolicy.DESTROY_AFTER_IDLE,
+                InstanceModel.deleted == False,
+                InstanceModel.job == None,  # noqa: E711
+            )
+            .options(joinedload(InstanceModel.project))
+        )
+        instances = res.scalars().all()
+
+        # TODO: need lock
+
+        for instance in instances:
+            last_time = instance.created_at.replace(tzinfo=datetime.timezone.utc)
+            if instance.last_job_processed_at is not None:
+                last_time = instance.last_job_processed_at.replace(tzinfo=datetime.timezone.utc)
+
+            current_time = get_current_datetime().replace(tzinfo=datetime.timezone.utc)
+
+            delta = datetime.timedelta(
+                seconds=parse_pretty_duration(instance.termination_idle_time)
```

Comment:
termination_idle_time should not be parsed at the time of processing. The db should should store already parsed termination_idle_time